core/system/src/unvic.c

```diff
@@ -280,163 +282,169 @@ int unvic_irq_level_get(void)
     }
 
-    /* gather information from current state */
-    src_id = g_active_box;
-    src_sp = svc_cx_validate_sf((uint32_t *) __get_PSP());
+    /* Source box: Get the current stack pointer. */
+    /* Note: We must use the current PSP as the SVCall can only give us the MSP
+     * register, since it was triggered from privileged mode. */
+    src_sp = context_validate_exc_sf(__get_PSP());
 
-    /* a proper context switch is only needed if changing box */
-    if(src_id != dst_id)
-    {
-        /* gather information from current state */
-        dst_sp = svc_cx_get_curr_sp(dst_id);
+    /* The stacked xPSR register clears the IRQn. */
+    xpsr = ipsr & ~0x1FF;
 
-        /* set the context stack pointer for the dst box */
-        *(__uvisor_config.uvisor_box_context) = g_svc_cx_context_ptr[dst_id];
+    /* The stacked return value is the second SVCall in the uVisor default ISR
+     * handler. */
+    unvic_thunk = (uint32_t) (svc_pc + 2);
 
-        /* switch boxes */
-        vmpu_switch(src_id, dst_id);
-    }
-    else
-    {
-        dst_sp = src_sp;
-    }
+    /* Forge a stack frame for the destination box. */
+    dst_sp = context_forge_exc_sf(src_sp, dst_id, dst_fn, unvic_thunk, xpsr, 0);
+
+    /* Perform the context switch-in to the destination box. */
+    /* This function halts if it finds an error. */
+    context_switch_in(CONTEXT_SWITCH_FUNCTION_ISR, dst_id, src_sp, dst_sp);
 
-    /* create unprivileged stack frame */
-    dst_sp_align = ((uint32_t) dst_sp & 0x4) ? 1 : 0;
-    dst_sp      -= (SVC_CX_EXC_SF_SIZE - dst_sp_align);
-
-    /* create stack frame for de-privileging the interrupt:
-     *   - all general purpose registers are ignored
-     *   - lr is the next SVCall
-     *   - the return address is the unprivileged handler
-     *   - xPSR is copied from the MSP
-     *   - pending IRQs are cleared in iPSR
-     *   - store alignment for future unstacking */
-    dst_sp[5] = ((uint32_t) svc_pc + 2) | 1;             /* lr */
-    dst_sp[6] = dst_fn;                                  /* return address */
-    dst_sp[7] = ipsr;                                    /* xPSR */
-    dst_sp[7] &= ~0x1FF;                                 /* IPSR - clear IRQn */
-    dst_sp[7] |= dst_sp_align << 9;                      /* xPSR - alignment */
-
-    /* save the current state */
-    svc_cx_push_state(src_id, TBOXCX_IRQ, src_sp, dst_id);
-    /* DEBUG_CX_SWITCH_IN(); */
-
-    /* de-privilege executionn */
-    __set_PSP((uint32_t) dst_sp);
+    /* De-privilege execution. */
     __set_CONTROL(__get_CONTROL() | 3);
 
-    /* DPRINTF("IRQ %d served with vector 0x%08X\n\r", irqn, */
-    /*                                                 g_unvic_vector[irqn]); */
-    /* FIXME add support for privacy (triggers register clearing) */
+    /* Return whether the destination box requires privacy or not. */
+    /* TODO: Context privacy is currently unsupported. */
```

Comment:
Should this remain as a FIXME instead of being turned into a TODO? I'm not sure if we differentiate between FIXME and TODO for the sake of tracking issues in code.